crates/bifrost/src/providers/replicated_loglet/remote_sequencer.rs

```diff
@@ -0,0 +1,554 @@
+    ///
+    /// It's up to the caller to retry on [`NetworkError`]
+    pub async fn send(
+        &self,
+        rpc_router: &RpcRouter<Append>,
+        sequencer: GenerationalNodeId,
+        msg: Append,
+    ) -> Result<RpcToken<Appended>, NetworkSendError<Append>> {
+        let outgoing = Outgoing::new(sequencer, msg).assign_connection(self.inner.clone());
+
+        rpc_router
+            .send_on_connection(outgoing)
+            .await
+            .map_err(|err| NetworkSendError::new(err.original.into_body(), err.source))
+    }
+
+    pub fn resolve_on_appended(
+        &self,
+        permit: OwnedSemaphorePermit,
+        rpc_token: RpcToken<Appended>,
+        commit_resolver: LogletCommitResolver,
+    ) {
+        let inflight_append = RemoteInflightAppend {
+            rpc_token,
+            commit_resolver,
+            permit,
+        };
+
+        if let Err(err) = self.tx.send(inflight_append) {
+            // if we failed to push this to be processed by the connection reactor task
+            // then we need to notify the caller
+            err.0
+                .commit_resolver
+                .error(AppendError::retryable(NetworkError::ConnectionClosed));
+        }
+    }
+
+    /// Handle all [`Appended`] responses
+    ///
+    /// This task will run until the [`AppendStream`] is dropped. Once dropped
+    /// all pending commits will be resolved with an error. it's up to the enqueuer
+    /// to retry if needed.
+    async fn handle_appended_responses(
+        known_global_tail: TailOffsetWatch,
+        connection: WeakConnection,
+        mut rx: mpsc::UnboundedReceiver<RemoteInflightAppend>,
+    ) -> anyhow::Result<()> {
+        let mut closed = std::pin::pin!(connection.closed());
+
+        // handle all rpc tokens in a loop, this loop will
+        // break only with a break reason.
+        // this reason can be one of the following:
+        // - Shutdown error
+        // - Terminal AppendError
+        // - Connection closed
+        let reason = loop {
+            let inflight = match rx.recv().await {
+                Some(inflight) => inflight,
+                None => {
+                    break BreakReason::Shutdown(ShutdownError);
+                }
+            };
+
+            tokio::select! {
+                appended = inflight.rpc_token.recv() => {
+                    let appended = match appended {
+                        Ok(appended) => appended,
+                        Err(err) => {
+                            break BreakReason::Shutdown(err);
+                        }
+                    };
+
+                    if let Err(err) = Self::resolve(&known_global_tail, inflight.commit_resolver, appended.into_body()) {
+                        break BreakReason::AppendError(err);
+                    }
+                },
+                _ = &mut closed => {
+                    break BreakReason::ConnectionClosed;
+                }
```

Comment:
What happens with `inflight` if we break out of the loop? Shouldn't we fail it like how we do it when draining `rx`?